From 50e1564ed4f68cdc7065104fa4aa2d5f0c6c07d3 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Mon, 28 Jul 2025 15:09:48 +0200
Subject: [PATCH] drm/stm: lvds: set clock rate

Added missing "set-rate" function. This function is needed if lower
software layers did not start or used a different LVDS clock value.

Change-Id: I01ba2717c40c7e36adaa998a1ddb9fa8e7070bcb
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/479210
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/stm/lvds.c | 80 ++++++++++++++++++++++++--------------
 1 file changed, 51 insertions(+), 29 deletions(-)

diff --git a/drivers/gpu/drm/stm/lvds.c b/drivers/gpu/drm/stm/lvds.c
index 08499c48b277..5d95c79ef80b 100644
--- a/drivers/gpu/drm/stm/lvds.c
+++ b/drivers/gpu/drm/stm/lvds.c
@@ -667,55 +667,49 @@ static unsigned long lvds_pixel_clk_recalc_rate(struct clk_hw *hw,
 	drm_dbg(drm, "ndiv %d , bdiv %d, mdiv %d, pll_out_khz %d\n",
 		ndiv, bdiv, mdiv, pll_out_khz);
 
+	clk_disable_unprepare(lvds->pclk);
+
 	/*
 	 * 1/7 because for each pixel in 1 lane there is 7 bits
 	 * We want pixclk, not bitclk
 	 */
-	lvds->pixel_clock_rate = pll_out_khz * 1000 * multiplier / 7;
-
-	clk_disable_unprepare(lvds->pclk);
-
-	return (unsigned long)lvds->pixel_clock_rate;
+	return (unsigned long)pll_out_khz * 1000 * multiplier / 7;
 }
 
 static long lvds_pixel_clk_round_rate(struct clk_hw *hw, unsigned long rate,
 				      unsigned long *parent_rate)
 {
 	struct stm_lvds *lvds = container_of(hw, struct stm_lvds, lvds_ck_px);
-	unsigned int pll_in_khz, bdiv = 0, mdiv = 0, ndiv = 0;
-	const struct drm_connector *connector;
-	const struct drm_display_mode *mode;
+	unsigned int pll_in_khz, pll_out_khz, bdiv = 0, mdiv = 0, ndiv = 0;
 	int multiplier;
 
-	connector = &lvds->connector;
-	if (!connector)
-		return -EINVAL;
-
-	if (list_empty(&connector->modes)) {
-		drm_dbg(connector->dev, "connector: empty modes list\n");
-		return -EINVAL;
-	}
-
-	mode = list_first_entry(&connector->modes,
-				struct drm_display_mode, head);
-
-	pll_in_khz = (unsigned int)(*parent_rate / 1000);
-
 	if (lvds_is_dual_link(lvds->link_type))
 		multiplier = 2;
 	else
 		multiplier = 1;
 
-	lvds_pll_get_params(lvds, pll_in_khz, mode->clock * 7 / multiplier, &bdiv, &mdiv, &ndiv);
+	pll_in_khz = (unsigned int)(*parent_rate / 1000);
+	pll_out_khz = rate * 7 / (1000 * multiplier),
+
+	lvds_pll_get_params(lvds, pll_in_khz, pll_out_khz, &bdiv, &mdiv, &ndiv);
 
 	/*
 	 * 1/7 because for each pixel in 1 lane there is 7 bits
 	 * We want pixclk, not bitclk
 	 */
-	lvds->pixel_clock_rate = (unsigned long)pll_get_clkout_khz(pll_in_khz, bdiv, mdiv, ndiv)
-					 * 1000 * multiplier / 7;
+	return (unsigned long)pll_get_clkout_khz(pll_in_khz, bdiv, mdiv, ndiv)
+	       * 1000 * multiplier / 7;
+}
+
+static int lvds_pixel_clk_set_rate(struct clk_hw *hw, unsigned long rate,
+				   unsigned long parent_rate)
+{
+	struct stm_lvds *lvds = container_of(hw, struct stm_lvds, lvds_ck_px);
 
-	return lvds->pixel_clock_rate;
+	/* Save pixel clock rate */
+	lvds->pixel_clock_rate = lvds_pixel_clk_round_rate(hw, rate, &parent_rate);
+
+	return 0;
 }
 
 static const struct clk_ops lvds_pixel_clk_ops = {
@@ -723,6 +717,7 @@ static const struct clk_ops lvds_pixel_clk_ops = {
 	.disable = lvds_pixel_clk_disable,
 	.recalc_rate = lvds_pixel_clk_recalc_rate,
 	.round_rate = lvds_pixel_clk_round_rate,
+	.set_rate = lvds_pixel_clk_set_rate,
 };
 
 static const struct clk_init_data clk_data = {
@@ -744,12 +739,39 @@ static void lvds_pixel_clk_unregister(void *data)
 static int lvds_pixel_clk_register(struct stm_lvds *lvds)
 {
 	struct device_node *node = lvds->dev->of_node;
-	int ret;
+	unsigned int pll_in_khz, bdiv, mdiv, ndiv;
+	int ret, multiplier, pll_out_khz;
+	struct clk_hw *hwclk;
+	u32 val;
 
 	lvds->lvds_ck_px.init = &clk_data;
 
-	/* set the rate by default at 148500000 */
-	lvds->pixel_clock_rate = 148500000;
+	if (lvds_is_dual_link(lvds->link_type))
+		multiplier = 2;
+	else
+		multiplier = 1;
+
+	val = lvds_read(lvds, lvds->primary->base + lvds->primary->ofs.PLLCR2);
+
+	ndiv = (val & PHY_PLLCR2_NDIV) >> 16;
+	bdiv = (val & PHY_PLLCR2_BDIV) >> 0;
+	mdiv = (unsigned int)lvds_read(lvds,
+				       lvds->primary->base + lvds->primary->ofs.PLLSDCR1);
+
+	hwclk = __clk_get_hw(lvds->pllref_clk);
+	if (hwclk) {
+		pll_in_khz = clk_hw_get_rate(hwclk) / 1000;
+		pll_out_khz = pll_get_clkout_khz(pll_in_khz, bdiv, mdiv, ndiv);
+
+		/*
+		 * 1/7 because for each pixel in 1 lane there is 7 bits
+		 * We want pixclk, not bitclk
+		 */
+		lvds->pixel_clock_rate = (unsigned long)pll_out_khz * 1000 * multiplier / 7;
+	} else {
+		/* set the rate by default at 148500000 */
+		lvds->pixel_clock_rate = 148500000;
+	}
 
 	ret = clk_hw_register(lvds->dev, &lvds->lvds_ck_px);
 	if (ret)
-- 
2.53.0

